### ISSUE TITLE
Strange behavior of 'label2rgb' function

### ISSUE THREAD

**Issue by simflin** [scikit-image/scikit-image#2674]
Hello
I've found some strange behavior of function `skimage.color.label2rgb` that I would consider as a bug. I usually use this function for coloring results of pixel-wise semantic segmentation methods. This usecase assumes that I have a predefined and fixed label-to-color correspondence and want to transform labels to colors according to this. And this is how `label2rgb` worked in 0.12.3. But in 0.13 if the function takes array of labels where some label 'k' is absent, it will produce result where label 'k+1' has the color of label 'k'.
Small code example for reproducing:
```import numpy as np
import skimage.color
labels = np.array([0, 2])
colors = np.array([[1,0,0], [0,1,0], [0,0,1]], dtype=np.float32)
print skimage.color.label2rgb(labels, colors=colors)
```
It will print out `[[1,0,0], [0,1,0]]` as the result in 0.13, and `[[1,0,0], [0,0,1]]` in 0.12.3. The last one is what I would expect to get in this case.
Could you please check if it is an expected behavior or it is a bug?

**Comment by jni** [comment]
Hi @simflin, thanks for the report!

It's unclear whether it's a bug or a feature. The behaviour changed in ea40a555, because of a report that `label2rgb` couldn't handle negative labels. Whoops!

We'll probably have to have a discussion about adding a keyword argument to control this behaviour, but in the meantime, your problem is actually the easiest possible case for label2rgb, and you could write your own one-liner:

```python
def label2rgb_direct(labels, colors):
    return colors[labels.ravel()].reshape(labels.shape + (3,))
```

You could also remap your colors to be sequential also:

```python
from skimage.segmentation import relabel_sequential
_, _, inv = relabel_sequential(labels)
new_colors = colors[inv]
```

**Comment by jni** [comment]
Actually, silly me, you don't even need the one liner:

```python
colored = colors[labels]
```

=P

**Comment by soupault** [comment]
>It's unclear whether it's a bug or a feature. The behaviour changed in ea40a55, because of a report that label2rgb couldn't handle negative labels. Whoops!

I agree. Negative labels and convenient label<->color correspondence aren't really compatible as is. The only non-destructive way to address the issue I see is to add `labels` argument where user is supposed to pass the list of labels of the same length as `colors`.

**Comment by simflin** [comment]
@soupault could you please briefly describe, what was the usecase for negative labels? It's not clear from code and I can't find any issue about it. Maybe there is some way to support this functionality without breaking backward compatibility.

**Comment by soupault** [comment]
@simflin seems to be a flash fix without a thorough discussion (most of which is in https://github.com/scikit-image/scikit-image/issues/2225). In my opinion, this should've taken another road, so I'd vote to revert the change.

Maybe we could get some commentary from @stefanv @ahojnnes ...

**Comment by simflin** [comment]
Well, seems if it is necessary to save the behavior checked by '[test_nonconsecutive](https://github.com/YueHazelZheng/scikit-image/blob/master/skimage/color/tests/test_colorlabel.py#L80)', an additional argument will have to be added since this behavior is incompatible with strict label-to-color correspondence. As for me, the result of label2rgb in 0.12.3 for this case was valid because docs explicitly say: "If the number of labels exceeds the number of colors, then the colors are cycled.". Problem with background color for negative labels is a bug, no doubt, but it can be fixed without impact on non-negative labels processing.

**Comment by jni** [comment]
My perspective on this is that this function (and all scikit-image functions) should primarily support use cases that are not solved by a single NumPy expression. It's annoying that we broke backwards compatibility, but the function was always designed to cycle colors, rather than provide a direct and reproducible mapping. This is vaguely implicit in the docstring: "The result of blending a cycling colormap (`colors`) for **each distinct value** in `label` with the image, at a certain alpha value."

**If** we decide to support the direct-mapping use case, my suggestion would be to add a `assume_consecutive_labels=False` argument (or somesuch) that would turn off all of the relabeling logic. But I'm a little reluctant to clutter an already cluttered interface.

@simflin do you consider my one-liner too arcane to be readable, or...?

@soupault "should've taken another road"... Care to specify? ;)

**Comment by stefanv** [comment]
The docstring does not specify explicitly what should happen in this case, so we have the option to go either way.  But by default we only have 10 colors, and you typically have many more labels.  In your use case you seem to have a one-to-one mapping, which we could support by also allowing a callable, `index_to_color(idx)` to be given for the `colors` argument; that way, you can make arbitrary mappings.

**Comment by rmegret** [comment]
I have two main issues with the new behavior:
1. It does not support reproducible display, as the color will depend on the content of the data. If you want to compare two label images, they may end up with different colors for the same label. Not good for data analysis such as plotting a scatter plot and a label image side by side.
2. It is surprising: `label2rgb(label, colors=['yellow', 'red', 'blue'])` should do simple and predictable: 0 is yellow, 1 is red and so on. With `label2rgb(label, colors=['yellow', 'red', 'blue'], bg_label=0)`, this is just shifted: 1 is yellow, 2 is red... With the new behavior, it depends on the content of `label` image. Sure, we can always RTFM, but a lot of people are going to get bitten by that given the apparent simplicity of the API.

I propose the labels are simply wrapped around using _modulo indexing_, which is one way to perform color cycling. In previous example, label=3 and label=-3 would cycle back to yellow. I guess there could be a discussion on how bg_label=-1 should impact the negative labels exactly (I think bg_label=None should produce pure modulo indexing, whereas bg_label=-1 should shift all negative labels by 1). This approach is reproducible, follow least-surprise principle, and addresses the cases of negative labels as well as labels larger than the number of colors.

@stefanv Having non consecutive labels mapped to consecutive colors seems to me a complex scenario. I think in such case it is the responsibility of the user to opt-in for fancy relabeling behavior if modulo cycling is not enough for them rather than making the simple use cases convoluted. The option could be called `relabel_sequential=True` (direct mapping is simple->simple call, non sequential relabeling is complex->opt-in with one option)

@jni Thanks for the one liner. The power of label2rgb is that is both shows the intent of the programmer, and has a combination of functionalities such as bg_label/bg_color and automatic conversion from named color lists (see my example) that have no equivalent in pure NumPy. I would not ditch the 'reproducible mapping' as this quite a big issue for scientific applications of skimage.

**Comment by jni** [comment]
@simflin @rmegret I'd like to invite your comments in #3015!
